File: java/testsuite/Jogl4Examples/PolygonClippingExample/src/options/CommandLineOptions.java
package options;

import model.PolygonClippingDebuggerModel;
import model.PolygonClippingFixtures;
import model.PolygonSurfaceTessellationMode;

public class CommandLineOptions
{
    private final PolygonClippingDebuggerModel model;
    private boolean offlineMode;

    public CommandLineOptions(PolygonClippingDebuggerModel model)
    {
        if ( model == null ) {
            throw new IllegalArgumentException("model can not be null");
        }
        this.model = model;
    }

    public void parse(String[] args)
    {
        if ( args == null ) {
            return;
        }

        for ( int i = 0; i < args.length; i++ ) {
            String arg = args[i];

            if ( "--fixture".equals(arg) || "--fixture-index".equals(arg) ) {
                ensureValue(args, i, arg);
                applyFixtureIndex(arg, args[++i]);
            }
            else if ( arg.startsWith("--fixture=") ) {
                applyFixtureIndex(arg, arg.substring("--fixture=".length()));
            }
            else if ( arg.startsWith("--fixture-index=") ) {
                applyFixtureIndex(arg, arg.substring("--fixture-index=".length()));
            }
            else if ( "--wires".equals(arg) ) {
                ensureValue(args, i, arg);
                model.getQuality().setWires(parseBoolean(args[++i], arg));
            }
            else if ( arg.startsWith("--wires=") ) {
                model.getQuality().setWires(parseBoolean(arg.substring("--wires=".length()), arg));
            }
            else if ( "--surfaces".equals(arg) ) {
                ensureValue(args, i, arg);
                model.getQuality().setSurfaces(parseBoolean(args[++i], arg));
            }
            else if ( arg.startsWith("--surfaces=") ) {
                model.getQuality().setSurfaces(parseBoolean(arg.substring("--surfaces=".length()), arg));
            }
            else if ( "--points".equals(arg) ) {
                ensureValue(args, i, arg);
                model.getQuality().setPoints(parseBoolean(args[++i], arg));
            }
            else if ( arg.startsWith("--points=") ) {
                model.getQuality().setPoints(parseBoolean(arg.substring("--points=".length()), arg));
            }
            else if ( "--tessellation-mode".equals(arg)
                      || "--polygon-surface-tessellation-mode".equals(arg) ) {
                ensureValue(args, i, arg);
                model.setPolygonSurfaceTessellationMode(
                    parseTessellationMode(args[++i], arg));
            }
            else if ( arg.startsWith("--tessellation-mode=") ) {
                model.setPolygonSurfaceTessellationMode(
                    parseTessellationMode(
                        arg.substring("--tessellation-mode=".length()), arg));
            }
            else if ( arg.startsWith("--polygon-surface-tessellation-mode=") ) {
                model.setPolygonSurfaceTessellationMode(
                    parseTessellationMode(
                        arg.substring("--polygon-surface-tessellation-mode=".length()), arg));
            }
            else if ( "--help".equals(arg) || "-h".equals(arg) ) {
                throw new IllegalArgumentException(usage());
            }
            else if ( "-offline".equals(arg) || "--offline".equals(arg) ) {
                offlineMode = true;
            }
            else {
                throw new IllegalArgumentException("Unknown argument: " + arg);
            }
        }
    }

    public boolean isOfflineMode()
    {
        return offlineMode;
    }

    public static String usage()
    {
        StringBuilder out = new StringBuilder();
        out.append("Usage: [--fixture <index|name>] [--fixture-index <0-based-index>] ")
            .append("[--wires <true|false|on|off>] [--surfaces <true|false|on|off>] ")
            .append("[--points <true|false|on|off>] ")
            .append("[--tessellation-mode <GLU|MONOTONE_DECOMPOSITION>] ")
            .append("[-offline]\n")
            .append("Alias: --polygon-surface-tessellation-mode <...>\n")
            .append("Examples:\n")
            .append("  --fixture 1 --wires on --surfaces off\n")
            .append("  --fixture TRIANGLE_VS_QUAD --surfaces true\n")
            .append("  --fixture-index 0 --wires false --surfaces true\n")
            .append("  --tessellation-mode MONOTONE_DECOMPOSITION\n")
            .append("  --tessellation-mode MONOTONE_DECOMPOSITION --wires on --surfaces on -offline\n")
            .append("  --tessellation-mode MONOTONE_DECOMPOSITION --points on -offline\n")
            .append("  -offline\n")
            .append("Available fixtures:\n");

        for ( int i = 0; i < PolygonClippingFixtures.CASES.length; i++ ) {
            out.append("  ")
                .append(i)
                .append(": ")
                .append(PolygonClippingFixtures.CASES[i].name())
                .append('\n');
        }

        return out.toString().stripTrailing();
    }

    private static PolygonSurfaceTessellationMode parseTessellationMode(
        String rawValue, String optionName)
    {
        if ( rawValue == null ) {
            throw new IllegalArgumentException(
                "Invalid tessellation mode for " + optionName + ": null");
        }

        String normalized = rawValue.trim().replace('-', '_').replace(' ', '_');
        for ( PolygonSurfaceTessellationMode mode
              : PolygonSurfaceTessellationMode.values() ) {
            if ( mode.name().equalsIgnoreCase(normalized)
                 || mode.name().equalsIgnoreCase(rawValue.trim())
                 || mode.getDisplayName().equalsIgnoreCase(rawValue.trim()) ) {
                return mode;
            }
        }

        throw new IllegalArgumentException(
            "Invalid tessellation mode for " + optionName + ": " + rawValue
                + " (expected GLU or MONOTONE_DECOMPOSITION)");
    }

    private void applyFixtureIndex(String optionName, String rawValue)
    {
        if ( "--fixture-index".equals(optionName) ) {
            applyFixtureIndexByNumber(rawValue, false);
            return;
        }

        if ( tryApplyFixtureIndexByNumber(rawValue, true) ) {
            return;
        }

        int fixtureIndex = findFixtureIndexByName(rawValue);
        if ( fixtureIndex < 0 ) {
            throw new IllegalArgumentException(
                "Unknown fixture name: " + rawValue);
        }

        applyFixtureIndex(fixtureIndex);
    }

    private void applyFixtureIndexByNumber(String rawValue, boolean oneBased)
    {
        int fixtureIndex = parseInt(rawValue, "--fixture-index");
        if ( oneBased ) {
            fixtureIndex--;
        }

        int totalFixtures = PolygonClippingFixtures.CASES.length;
        if ( fixtureIndex < 0 || fixtureIndex >= totalFixtures ) {
            throw new IllegalArgumentException(
                "Fixture index out of range: " + rawValue
                    + " (valid range: " + ( oneBased ? "1" : "0" )
                    + ".." + ( oneBased ? totalFixtures : (totalFixtures - 1) )
                    + ")");
        }

        applyFixtureIndex(fixtureIndex);
    }

    private boolean tryApplyFixtureIndexByNumber(String rawValue, boolean oneBased)
    {
        try {
            applyFixtureIndexByNumber(rawValue, oneBased);
            return true;
        }
        catch ( IllegalArgumentException e ) {
            if ( oneBased && !looksNumeric(rawValue) ) {
                return false;
            }
            throw e;
        }
    }

    private void applyFixtureIndex(int fixtureIndex)
    {
        model.setTestIndex(fixtureIndex);
        System.out.println("[PolygonClippingExample] Selected fixture "
            + fixtureIndex + " [" + PolygonClippingFixtures.CASES[fixtureIndex].name() + "]");
    }

    private static int findFixtureIndexByName(String rawValue)
    {
        if ( rawValue == null ) {
            return -1;
        }

        for ( int i = 0; i < PolygonClippingFixtures.CASES.length; i++ ) {
            if ( PolygonClippingFixtures.CASES[i].name().equalsIgnoreCase(rawValue) ) {
                return i;
            }
        }
        return -1;
    }

    private static void ensureValue(String[] args, int index, String optionName)
    {
        if ( index + 1 >= args.length ) {
            throw new IllegalArgumentException("Missing value for " + optionName);
        }
    }

    private static int parseInt(String rawValue, String optionName)
    {
        try {
            return Integer.parseInt(rawValue);
        }
        catch ( NumberFormatException e ) {
            throw new IllegalArgumentException(
                "Invalid integer for " + optionName + ": " + rawValue);
        }
    }

    private static boolean looksNumeric(String rawValue)
    {
        if ( rawValue == null || rawValue.isBlank() ) {
            return false;
        }

        int start = rawValue.charAt(0) == '-' ? 1 : 0;
        if ( start >= rawValue.length() ) {
            return false;
        }

        for ( int i = start; i < rawValue.length(); i++ ) {
            if ( !Character.isDigit(rawValue.charAt(i)) ) {
                return false;
            }
        }
        return true;
    }

    private static boolean parseBoolean(String rawValue, String optionName)
    {
        String value = rawValue.trim().toLowerCase();
        if ( "true".equals(value) || "on".equals(value) || "1".equals(value)
             || "yes".equals(value) ) {
            return true;
        }
        if ( "false".equals(value) || "off".equals(value) || "0".equals(value)
             || "no".equals(value) ) {
            return false;
        }

        throw new IllegalArgumentException(
            "Invalid boolean for " + optionName + ": " + rawValue);
    }
}

File: cpp/testsuite/OpenGL4Examples/PolygonClippingExample/src/options/CommandLineOptions.cpp
#include "options/CommandLineOptions.h"

#include <cctype>
#include <cerrno>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <climits>

#include "model/PolygonClippingFixtures.h"

CommandLineOptions::CommandLineOptions()
    : offlineMode(false)
    , hasFixtureIndex(false)
    , fixtureIndex(0)
    , hasWires(false)
    , wires(true)
    , hasSurfaces(false)
    , surfaces(true)
    , hasPoints(false)
    , points(true)
    , hasTessellationMode(false)
    , tessellationMode(PolygonSurfaceTessellationMode::GLU)
{}

static bool strEqI(const char* a, const char* b)
{
    while (*a && *b) {
        if (std::tolower((unsigned char)*a) != std::tolower((unsigned char)*b)) return false;
        a++; b++;
    }
    return *a == '\0' && *b == '\0';
}

static bool parseBoolean(const char* raw, bool& out)
{
    if (strEqI(raw, "true") || strEqI(raw, "on") || strEqI(raw, "1") || strEqI(raw, "yes")) {
        out = true; return true;
    }
    if (strEqI(raw, "false") || strEqI(raw, "off") || strEqI(raw, "0") || strEqI(raw, "no")) {
        out = false; return true;
    }
    std::fprintf(stderr, "[PolygonClippingExample] Invalid boolean value: %s\n", raw);
    return false;
}

static bool looksNumeric(const char* raw)
{
    if (!raw || !*raw) return false;
    const char* p = raw;
    if (*p == '-') p++;
    if (!*p) return false;
    while (*p) {
        if (!std::isdigit((unsigned char)*p)) return false;
        p++;
    }
    return true;
}

static bool parseInt(const char* raw, int& out)
{
    if (!looksNumeric(raw)) {
        std::fprintf(stderr, "[PolygonClippingExample] Invalid integer: %s\n", raw);
        return false;
    }
    errno = 0;
    char* end = nullptr;
    long v = std::strtol(raw, &end, 10);
    if (errno != 0 || end == raw || *end != '\0' || v < INT_MIN || v > INT_MAX) {
        std::fprintf(stderr, "[PolygonClippingExample] Integer out of range: %s\n", raw);
        return false;
    }
    out = (int)v;
    return true;
}

static bool parseTessellationMode(const char* raw, PolygonSurfaceTessellationMode& out)
{
    if (strEqI(raw, "GLU")) {
        out = PolygonSurfaceTessellationMode::GLU;
        return true;
    }

    const char* normalized = raw;
    if (strEqI(raw, "MONOTONE_DECOMPOSITION")
        || strEqI(raw, "MONOTONE-DECOMPOSITION")
        || strEqI(raw, "Monotone decomposition")
        || strEqI(raw, "monotone_decomposition")) {
        out = PolygonSurfaceTessellationMode::MONOTONE_DECOMPOSITION;
        return true;
    }

    std::fprintf(stderr,
        "[PolygonClippingExample] Invalid tessellation mode: %s"
        " (expected GLU or MONOTONE_DECOMPOSITION)\n", raw);
    return false;
}

static int findFixtureIndexByName(const char* name)
{
    for (int i = 0; i < PolygonClippingFixtures::COUNT; i++) {
        if (strEqI(PolygonClippingFixtures::CASES[i].name, name)) return i;
    }
    return -1;
}

static bool applyFixtureByValue(const char* raw, bool oneBased, int& indexOut)
{
    if (looksNumeric(raw)) {
        int v = 0;
        if (!parseInt(raw, v)) return false;
        if (oneBased) v--;
        if (v < 0 || v >= PolygonClippingFixtures::COUNT) {
            std::fprintf(stderr,
                "[PolygonClippingExample] Fixture index out of range: %s\n", raw);
            return false;
        }
        indexOut = v;
        return true;
    }
    int idx = findFixtureIndexByName(raw);
    if (idx < 0) {
        std::fprintf(stderr, "[PolygonClippingExample] Unknown fixture name: %s\n", raw);
        return false;
    }
    indexOut = idx;
    return true;
}

static const char* startsWith(const char* str, const char* prefix)
{
    while (*prefix) {
        if (*str != *prefix) return nullptr;
        str++; prefix++;
    }
    return str;
}

bool CommandLineOptions::parse(int argc, char** argv, CommandLineOptions& out)
{
    out = CommandLineOptions();

    for (int i = 1; i < argc; i++) {
        const char* arg = argv[i];
        const char* val = nullptr;

        if (std::strcmp(arg, "--offline") == 0 || std::strcmp(arg, "-offline") == 0) {
            out.offlineMode = true;
            continue;
        }

        if (std::strcmp(arg, "--help") == 0 || std::strcmp(arg, "-h") == 0) {
            printUsage();
            return false;
        }

        if (std::strcmp(arg, "--fixture") == 0 || std::strcmp(arg, "--fixture-index") == 0) {
            if (i + 1 >= argc) {
                std::fprintf(stderr, "[PolygonClippingExample] Missing value for %s\n", arg);
                return false;
            }
            val = argv[++i];
            bool zeroBased = (std::strcmp(arg, "--fixture-index") == 0);
            out.hasFixtureIndex = applyFixtureByValue(val, !zeroBased, out.fixtureIndex);
            if (!out.hasFixtureIndex) return false;
            continue;
        }

        if ((val = startsWith(arg, "--fixture=")) != nullptr) {
            out.hasFixtureIndex = applyFixtureByValue(val, true, out.fixtureIndex);
            if (!out.hasFixtureIndex) return false;
            continue;
        }

        if ((val = startsWith(arg, "--fixture-index=")) != nullptr) {
            out.hasFixtureIndex = applyFixtureByValue(val, false, out.fixtureIndex);
            if (!out.hasFixtureIndex) return false;
            continue;
        }

        if (std::strcmp(arg, "--wires") == 0) {
            if (i + 1 >= argc) {
                std::fprintf(stderr, "[PolygonClippingExample] Missing value for --wires\n");
                return false;
            }
            out.hasWires = parseBoolean(argv[++i], out.wires);
            if (!out.hasWires) return false;
            continue;
        }

        if ((val = startsWith(arg, "--wires=")) != nullptr) {
            out.hasWires = parseBoolean(val, out.wires);
            if (!out.hasWires) return false;
            continue;
        }

        if (std::strcmp(arg, "--surfaces") == 0) {
            if (i + 1 >= argc) {
                std::fprintf(stderr, "[PolygonClippingExample] Missing value for --surfaces\n");
                return false;
            }
            out.hasSurfaces = parseBoolean(argv[++i], out.surfaces);
            if (!out.hasSurfaces) return false;
            continue;
        }

        if ((val = startsWith(arg, "--surfaces=")) != nullptr) {
            out.hasSurfaces = parseBoolean(val, out.surfaces);
            if (!out.hasSurfaces) return false;
            continue;
        }

        if (std::strcmp(arg, "--points") == 0) {
            if (i + 1 >= argc) {
                std::fprintf(stderr, "[PolygonClippingExample] Missing value for --points\n");
                return false;
            }
            out.hasPoints = parseBoolean(argv[++i], out.points);
            if (!out.hasPoints) return false;
            continue;
        }

        if ((val = startsWith(arg, "--points=")) != nullptr) {
            out.hasPoints = parseBoolean(val, out.points);
            if (!out.hasPoints) return false;
            continue;
        }

        if (std::strcmp(arg, "--tessellation-mode") == 0
            || std::strcmp(arg, "--polygon-surface-tessellation-mode") == 0) {
            if (i + 1 >= argc) {
                std::fprintf(stderr, "[PolygonClippingExample] Missing value for %s\n", arg);
                return false;
            }
            out.hasTessellationMode = parseTessellationMode(argv[++i], out.tessellationMode);
            if (!out.hasTessellationMode) return false;
            continue;
        }

        if ((val = startsWith(arg, "--tessellation-mode=")) != nullptr) {
            out.hasTessellationMode = parseTessellationMode(val, out.tessellationMode);
            if (!out.hasTessellationMode) return false;
            continue;
        }

        if ((val = startsWith(arg, "--polygon-surface-tessellation-mode=")) != nullptr) {
            out.hasTessellationMode = parseTessellationMode(val, out.tessellationMode);
            if (!out.hasTessellationMode) return false;
            continue;
        }

        std::fprintf(stderr, "[PolygonClippingExample] Unknown argument: %s\n", arg);
        return false;
    }

    return true;
}

void CommandLineOptions::printUsage()
{
    std::fprintf(stderr,
        "Usage: [--fixture <index|name>] [--fixture-index <0-based-index>]\n"
        "       [--wires <true|false|on|off>] [--surfaces <true|false|on|off>]\n"
        "       [--points <true|false|on|off>]\n"
        "       [--tessellation-mode <GLU|MONOTONE_DECOMPOSITION>]\n"
        "       [--offline]\n"
        "Alias: --polygon-surface-tessellation-mode <...>\n"
        "Examples:\n"
        "  --fixture 1 --wires on --surfaces off\n"
        "  --fixture TRIANGLE_VS_QUAD --surfaces true\n"
        "  --fixture-index 0 --wires false --surfaces true\n"
        "  --tessellation-mode MONOTONE_DECOMPOSITION\n"
        "  --tessellation-mode MONOTONE_DECOMPOSITION --wires on --surfaces on --offline\n"
        "  --tessellation-mode MONOTONE_DECOMPOSITION --points on --offline\n"
        "  --offline\n"
        "Available fixtures:\n");

    for (int i = 0; i < PolygonClippingFixtures::COUNT; i++) {
        std::fprintf(stderr, "  %d: %s\n", i, PolygonClippingFixtures::CASES[i].name);
    }
}
